snapcraft/parts/plugins/kernel_build.sh

```diff
@@ -0,0 +1,323 @@
+#!/bin/sh
+
+set -eu
+
+# TODO: should these be embedded these in the plugin?
+# Required kernel configs for Ubuntu Core
+_required_generic="
+DEVTMPFS
+DEVTMPFS_MOUNT
+TMPFS_POSIX_ACL
+IPV6
+SYSVIPC
+SYSVIPC_SYSCTL
+VFAT_FS
+NLS_CODEPAGE_437
+NLS_ISO8859_1"
+
+_required_security="
+SECURITY
+SECURITY_APPARMOR
+SYN_COOKIES
+STRICT_DEVMEM
+DEFAULT_SECURITY_APPARMOR
+SECCOMP
+SECCOMP_FILTER"
+
+_required_snappy="
+RD_LZMA
+KEYS
+ENCRYPTED_KEYS
+SQUASHFS
+SQUASHFS_XATTR
+SQUASHFS_XZ
+SQUASHFS_LZO"
+
+_required_systemd="
+DEVTMPFS
+CGROUPS
+INOTIFY_USER
+SIGNALFD
+TIMERFD
+EPOLL
+NET
+SYSFS
+PROC_FS
+FHANDLE
+BLK_DEV_BSG
+NET_NS
+IPV6
+AUTOFS4_FS
+TMPFS_POSIX_ACL
+TMPFS_XATTR
+SECCOMP"
+
+_required_boot="SQUASHFS"
+
+  kver="$(basename "${CRAFT_PART_INSTALL}/modules/"*)"
+
+for arg; do
+  case $arg in
+    defconfig=*)   defconfig=${arg#*=}   ;; # valid: list
+    flavour=*)     flavour=${arg#*=}     ;; # valid: string, trumps kdefconfig
+    configs=*)     configs=${arg#*=}     ;; # valid: list, trumps kconfig-flavour
+    enable-zfs=*)  enable_zfs=${arg#*=}  ;; # valid: list, true|false
+    enable-perf=*) enable_perf=${arg#*=} ;; # valid: list, true|false
+  esac
+done
+
+# Set of valid targets specified by plugin
+build_target="$KERNEL_IMAGE $KERNEL_TARGET"
+
+# Remove artifacts from prior builds
+cleanup() {
+  printf 'Cleaning previous build first...\n'
+
+  rm -rf "${CRAFT_PART_INSTALL}/modules"
+  unlink "${CRAFT_PART_INSTALL}/lib/modules"
+}
+
+# Create specified defconfig
+gen_defconfig() {
+  make -j1                    \
+       -C "$CRAFT_PART_SRC"   \
+        O="$CRAFT_PART_BUILD" \
+        "$defconfig"
+}
+
+# Update config with flavour
+gen_flavour_config() {
+  OLDPWD="$PWD"
+  ubuntuconfig="${CRAFT_PART_SRC}/CONFIGS/${CRAFT_ARCH_BUILD_FOR}-config.flavour.${flavour}"
+  cd "${CRAFT_PART_SRC}"
+
+  # Generate the configs
+  # Don't let failures stop us
+  fakeroot debian/rules clean genconfigs || true
+
+  cat "$ubuntuconfig" > "${CRAFT_PART_BUILD}/.config"
+
+  fakeroot debian/rules clean
+
+  rm -rf CONFIGS
+
+  cd "$OLDPWD"
+}
+
+# Add any specified kconfig overrides
+add_kconfigs() {
+  kconfigs="$1"
+  _kconfigs="$(echo "$kconfigs" | sed -e 's/,/\n/g')"
+
+  printf 'Applying extra config....'
+
+  # Need to echo the list twice to force all Kconfigs to pickup the change
```

Comment:
Part of me thinks this isn't true but I don't know what Kconfigs might cause this to test it.